src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,489 @@
+    *       <a href="https://en.wikipedia.org/wiki/Fluent_interface">fluent interface</a>.
+    */
+   public RepeatingTaskThread setFrequencyLimit(double frequencyLimit)
+   {
+      loopPeriodLowerLimit = Conversions.hertzToSeconds(frequencyLimit);
+      return this;
+   }
+
+   /**
+    * Removes any limit to the loop frequency that may have been set.
+    * Equivalent to calling {@code setFrequencyLimit(UNLIMITED_FREQUENCY)}.
+    */
+   public void removeFrequencyLimit()
+   {
+      setFrequencyLimit(UNLIMITED_FREQUENCY);
+   }
+
+   @Override
+   public void start()
+   {
+      state.setRunning(true);
+      super.start();
+   }
+
+   /**
+    * Signals the thread to start repeating indefinitely.
+    * If the thread has not been started yet, calling this
+    * method will also start the thread.
+    */
+   public void startRepeating()
+   {
+      if (!state.isRunning())
+         start();
+
+      setRepeating(true);
+   }
+
+   /**
+    * Signals the thread to stop repeating once the current repetition finishes.
+    * The thread will be paused until the value of remaining repetitions changes, or the thread is killed.
+    */
+   public void stopRepeating()
+   {
+      setRepeating(false);
+   }
+
+   /**
+    * Signal the thread to repeat indefinitely, or to stop repeating.
+    *
+    * @param repeating Whether the thread should be repeating.
+    */
+   public void setRepeating(boolean repeating)
+   {
+      if (repeating)
+         setRemaining(REPEAT_INDEFINITELY);
+      else
+         setRemaining(0);
+   }
+
+   /**
+    * Signal the thread to loop for the passed in number of repetitions.
+    * This overrides the remaining number of repetitions, regardless of its previous value.
+    * <p>
+    * This method also accepts {@link #REPEAT_INDEFINITELY}.
+    * <p>
+    * To add or subtract to the number of repetitions the thread should loop, use {@link #addRemaining(int)}.
+    *
+    * @param repetitions The number of repetitions the thread should loop after this call.
+    */
+   public void setRemaining(long repetitions)
+   {
+      state.setRemaining(repetitions);
+   }
+
+   /**
+    * Add N repetitions to the remaining repetition counter.
+    * If the thread was paused, adding repetitions begins the loop.
+    * <p>
+    * You may also subtract from the number of remaining repetitions by passing in a negative number.
+    * If the resulting number of repetitions is 0, the loop will be paused.
+    * This method cannot cause the remaining repetition count to go below 0.
+    * <p>
+    * This method does not do anything if the thread is repeating indefinitely.
+    *
+    * @param repetitions The number of repetitions to add. This can be a negative value for subtraction.
+    */
+   public void addRemaining(int repetitions)
+   {
+      state.addRemaining(repetitions);
+   }
+
+   /**
+    * Access the state of this thread.
+    *
+    * @return The state of this thread.
+    */
+   public State getRepetitionState()
```

Comment:
I think the user should not even know about the internal State class or interact with it.

Reply:
Done!